client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +291,62 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map(_.pipeBodyThrough(until(disposed)))
-      }
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message match {
+      case req: Request[F] =>
+        Resource.eval(Ref[F].of(false)).flatMap { disposed =>
+          val reqAugmented =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+          Resource
+            .eval(app(reqAugmented))
+            .onFinalize(disposed.set(true))
+            .flatMap(run(_, Option(disposed)))
+        }
+      case resp: Response[F] =>
+        for {
+          disposed <- Resource.eval(refOp.fold(Ref[F].of(false))(F.pure))
+          channel <- Resource.eval(Channel.synchronous[F, Chunk[Byte]])
+
+          _ <- resp.body.chunks
+            .through(channel.sendAll)
+            .compile
+            .drain
+            .background
+
+          ifDisposed = Stream
+            .eval(disposed.get)
+            .ifM(
+              Stream.eval(F.raiseError(new IOException("response was disposed"))),
+              Stream.empty,
+            )
+
+          r = resp.withBodyStream(
+            channel.stream.unchunks
+              .ifEmpty(ifDisposed)
+              .onFinalize(
+                channel.stream.compile.drain
+                  .guarantee(disposed.set(true))
+              )
+          )
+
+          _ <- Resource.onFinalize {
+            channel.close.void.guarantee {
+              disposed.get
+                .ifM(
+                  F.unit,
+                  r.body.compile.drain.void,
+                )
+                .guarantee(disposed.set(true))
+            }
```

Comment:
Wouldn't this close the `channel` possibly while it's still being written to, in the background? It's not obvious to me why we should ever close it ourselves: the only time it should be closed is when the `sendAll` completes.

Reply:
You're right. I removed it.